src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -0,0 +1,615 @@
+    return extent->insert(oc, key, value);
+  }).safe_then([this, oc, child_pt] (auto mresult) {
+    if (mresult.status == mutation_status_t::SUCCESS) {
+      return insert_ertr::make_ready_future<mutation_result_t>(mresult);
+    } else if (mresult.status == mutation_status_t::SPLITTED) {
+      return handle_split(oc, child_pt, mresult);
+    } else {
+     return insert_ret(
+            insert_ertr::ready_future_marker{},
+            mutation_result_t(mutation_status_t::SUCCESS, std::nullopt, std::nullopt));
+    }
+  });
+}
+
+OMapInnerNode::rm_key_ret
+OMapInnerNode::rm_key(omap_context_t oc, const std::string &key)
+{
+  logger().debug("{}: {}","OMapInnerNode",  __func__);
+  auto child_pt = get_containing_child(key);
+  auto laddr = child_pt->get_node_key().laddr;
+  return omap_load_extent(oc, laddr, get_meta().depth - 1).safe_then(
+    [this, oc, &key, child_pt] (auto extent) {
+    return extent->rm_key(oc, key)
+      .safe_then([this, oc, child_pt, extent = std::move(extent)] (auto mresult) {
+      if (mresult.status == mutation_status_t::SUCCESS ||
+          mresult.status == mutation_status_t::FAIL) {
+        return rm_key_ertr::make_ready_future<mutation_result_t>(mresult);
+      } else if (mresult.status == mutation_status_t::NEED_MERGE) {
+        if (get_node_size() >1)
+          return merge_entry(oc, child_pt, *(mresult.need_merge));
+        else
+          return rm_key_ret(
+                 rm_key_ertr::ready_future_marker{},
+                 mutation_result_t(mutation_status_t::SUCCESS,
+                                   std::nullopt, std::nullopt));
+      } else if (mresult.status == mutation_status_t::SPLITTED) {
+        return handle_split(oc, child_pt, mresult);
+      } else {
+        return rm_key_ertr::make_ready_future<mutation_result_t>(mresult);
+      }
+    });
+  });
+}
+
+OMapInnerNode::list_keys_ret
+OMapInnerNode::list_keys(omap_context_t oc, std::string &start, size_t max_result_size)
+{
+  logger().debug("{}: {}","OMapInnerNode",  __func__);
+  auto  child_iter = get_containing_child(start);
+
+  return seastar::do_with(child_iter, iter_end(), list_keys_result_t(), [=, &start]
+    (auto &biter, auto &eiter, auto &result) {
+    result.next = start;
+    return crimson::do_until([=, &biter, &eiter, &result] ()
+       -> list_keys_ertr::future<bool> {
+      if (biter == eiter  || result.keys.size() == max_result_size)
+        return list_keys_ertr::make_ready_future<bool>(true);
+
+      auto laddr = biter->get_node_key().laddr;
+      return omap_load_extent(oc, laddr, get_meta().depth - 1).safe_then(
+        [=, &biter, &eiter, &result] (auto &&extent) {
+        return extent->list_keys(oc, result.next, max_result_size - result.keys.size())
+          .safe_then([&biter, &eiter, &result] (auto &&list) mutable {
+          if (!list.keys.empty())
+            result.keys.insert(result.keys.end(), list.keys.begin(),list.keys.end());
```

Comment:
what is the type of result.keys()?  looks I have to checkout to tell;  I'm worried about finding a std::list